src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -317,6 +328,20 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
         None
     }
 
+    override protected def contains0(key: A, hash: Int, level: Int): Boolean = {
+      // Note: this code is duplicated with `get0`
+      val index = (hash >>> level) & 0x1f
+      val mask = (1 << index)
+      if (bitmap == - 1) {
+        elems(index & 0x1f).contains0(key, hash, level + 5)
+      } else if ((bitmap & mask) != 0) {
+        val offset = Integer.bitCount(bitmap & (mask-1))
+        elems(offset).contains0(key, hash, level + 5)
```

Comment:
If I understood it correctly, this traverses down the tree, calculating the 32 bit digits of the hash.
In Vector this path is unrolled to avoid all those method calls. While I think it's a very excessive optimization, it might be worth trying it here also.